src/sparse/linalg/ordering.rs

```diff
@@ -13,74 +13,223 @@ pub struct Ordering<I> {
+            // We found a new connected component. This number will be reverse-inverted later.
+            connected_parts.push(perm_index);
+            // Employ the George-Liu pseudoperipheral vertex finder to find a new starting vertex.
+            find_pseudoperipheral_vertex(&visited, &degrees, &mat).unwrap()
+        });
 
-    while perm.len() < nb_vertices {
-        // find the non-visited vertex with the lowest degree
-        let mut min_deg = nb_vertices;
-        let mut min_deg_vert = 0;
-        for (vert, vis) in visited.iter().enumerate() {
-            let vert_deg = degrees[vert];
-            if !vis && vert_deg <= min_deg {
-                min_deg = vert_deg;
-                min_deg_vert = vert;
+        // Write the next permutation in reverse order.
+        perm[nb_vertices - perm_index - 1] = I::from_usize(current_vertex);
+        visited[current_vertex.index()] = true;
+
+        // Find, sort, and push all new neighbors of the current vertex.
+        let outer = mat.outer_view(current_vertex.index()).unwrap();
+        neighbors.clear();
+        for &neighbor in outer.indices() {
+            if !visited[neighbor.index()] {
+                neighbors.push((degrees[neighbor.index()], neighbor));
+                visited[neighbor.index()] = true;
             }
         }
-        deque.clear();
-        deque.push_back(min_deg_vert);
 
-        while let Some(cur_vert) = deque.pop_front() {
-            if visited[cur_vert] {
-                continue;
-            }
-            perm.push(I::from_usize(cur_vert));
-            visited[cur_vert.index()] = true;
-            let outer = mat.outer_view(cur_vert.index()).unwrap();
-            neighbors.clear();
-            for &neighbor in outer.indices() {
-                if !visited[neighbor.index()] {
-                    neighbors.push((degrees[neighbor.index()], neighbor));
-                }
-            }
-            neighbors.sort_by_key(|&(deg, _)| deg);
-            for (_deg, neighbor) in &neighbors {
-                deque.push_back(neighbor.index());
-            }
+        // Missed optimization: match small sizes explicitly, sort using sorting networks.
+        // This especially makes sense if swaps are predictably compiled into cmov instructions,
+        // which they aren't currently, see https://github.com/rust-lang/rust/issues/53823.
+        // For more information on how to do sorting networks efficiently see https://arxiv.org/pdf/1505.01962.pdf.
+        neighbors.sort_unstable_by_key(|&(deg, _)| deg);
+
+        for (_deg, neighbor) in &neighbors {
+            deque.push_back(neighbor.index());
         }
-        connected_parts.push(perm.len());
     }
 
+    connected_parts.push(nb_vertices);
+
     Ordering {
         perm: PermOwnedI::new(perm),
-        connected_parts,
+        connected_parts: connected_parts
+            .into_iter()
+            .map(|i| nb_vertices - i)
+            .rev()
+            .collect(),
```

Comment:
I'm wondering if this could not be done in place here, something like:

```rust
conected_parts.iter_mut().map(|i| *i = nb_vertices - *i);
coneccted_parts.reverse();
```

Reply:
Yes, of course. I made this change, replacing the `map` with a `for_each` though ;-)